llama-index-core/llama_index/core/indices/property_graph/transformations/schema_llm.py

```diff
@@ -0,0 +1,327 @@
+            entity_cls = create_model(
+                "Entity",
+                type=(
+                    possible_entities if strict else str,
+                    Field(
+                        ...,
+                        description=(
+                            "Entity in a knowledge graph. Only extract entities with types that are listed as valid: "
+                            + str(possible_entities)
+                        ),
+                    ),
+                ),
+                name=(str, ...),
+            )
+
+            possible_relations = possible_relations or DEFAULT_RELATIONS
+            relation_cls = create_model(
+                "Relation",
+                type=(
+                    possible_relations if strict else str,
+                    Field(
+                        ...,
+                        description=(
+                            "Relation in a knowledge graph. Only extract relations with types that are listed as valid: "
+                            + str(possible_relations)
+                        ),
+                    ),
+                ),
+            )
+
+            triplet_cls = create_model(
+                "Triplet",
+                subject=(entity_cls, ...),
+                relation=(relation_cls, ...),
+                object=(entity_cls, ...),
+            )
+
+            def validate(v: Any, values: Any) -> Any:
+                """Validate triplets."""
+                passing_triplets = []
+                for i, triplet in enumerate(v):
+                    # cleanup
+                    try:
+                        for key in triplet:
+                            triplet[key]["type"] = triplet[key]["type"].replace(
+                                " ", "_"
+                            )
+                            triplet[key]["type"] = triplet[key]["type"].upper()
+
+                        # validate, skip if invalid
+                        _ = triplet_cls(**triplet)
+                        passing_triplets.append(v[i])
+                    except (KeyError, ValueError):
+                        continue
+
+                return passing_triplets
+
+            root = validator("triplets", pre=True)(validate)
+            kg_schema_cls = create_model(
+                "KGSchema",
+                __validators__={"validator1": root},
+                triplets=(List[triplet_cls], ...),
+            )
+            kg_schema_cls.__doc__ = "Knowledge Graph Schema."
+
+        super().__init__(
+            llm=llm,
+            extract_prompt=extract_prompt or DEFAULT_SCHEMA_PATH_EXTRACT_PROMPT,
+            kg_schema_cls=kg_schema_cls,
+            kg_validation_schema=kg_validation_schema or DEFAULT_VALIDATION_SCHEMA,
+            num_workers=num_workers,
+            max_triplets_per_chunk=max_triplets_per_chunk,
+            strict=strict,
+        )
+
+    @classmethod
+    def class_name(cls) -> str:
+        return "ExtractTripletsFromText"
```

Comment:
If I'm not mistaken, this should return `"SchemaLLMPathExtractor"`?

Reply:
Good catch. So many name iterations

Reply:
There's one thing I have been unsure about. Why do we need to define so many `class_name()` in classes and their subclasses instead of just define once in the (base) classes?

For example:

```python
>>> class Base:
...     @classmethod
...     def class_name(cls):
...         return cls.__name__
... 
>>> class A(Base):
...     pass
... 
>>> class B(Base):
...     pass
... 
>>> Base.class_name()
'Base'
>>> A.class_name()
'A'
>>> B.class_name()
'B'
>>> Base().class_name()
'Base'
>>> A().class_name()
'A'
>>> B().class_name()
'B'
>>> 
```

Reply:
Its for serialization and loading of pyndatic components

`pydantic_obj.dict()` has no details on the actual class that it was. With a dedicated identifier, it makes it easier. (This is mainly for llamacloud)

If we used `cls.__name__` and the class name ever changes, then it would break any stored objects

Reply:
Got it, thanks for the explanation!